airbyte-integrations/connectors/source-braintree/source_braintree/streams.py

```diff
@@ -0,0 +1,215 @@
+        """
+
+    @abstractmethod
+    def get_items(self, start_date: datetime):
+        """
+        braintree SDK gateway object for items list
+        """
+
+    def get_json_schema(self):
+        return self.model.schema()
+
+    def stream_slices(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+
+        current_datetime = pendulum.utcnow()
+        if sync_mode == SyncMode.full_refresh:
+            return [{self.cursor_field or "start_date": self._start_date or current_datetime}]
+        stream_state_start_date = stream_state.get(self.cursor_field)
+        if stream_state_start_date:
+            stream_state_start_date = pendulum.parse(stream_state_start_date)
+        start_date = stream_state_start_date or self._start_date or current_datetime
+        return [{self.cursor_field: start_date}]
+
+    def get_updated_state(
+        self,
+        current_stream_state: Mapping[str, Any],
+        latest_record: Mapping[str, Any],
+    ):
+        next_state = latest_record.get(self.cursor_field)
+        current_state = current_stream_state.get(self.cursor_field)
+        current_state = pendulum.parse(current_state) if current_state else next_state
+        return {self.cursor_field: max(current_state, next_state).strftime("%Y-%m-%d %H:%M:%S")}
+
+    @staticmethod
+    def get_json_from_resource(resource_obj: Union[AttributeGetter, List[AttributeGetter]]):
+
+        if isinstance(resource_obj, list):
+            return [obj if not isinstance(obj, AttributeGetter) else BraintreeStream.get_json_from_resource(obj) for obj in resource_obj]
+        obj_dict = resource_obj.__dict__
+        return {
+            attr: obj_dict[attr]
+            if not isinstance(obj_dict[attr], (AttributeGetter, list))
+            else BraintreeStream.get_json_from_resource(obj_dict[attr])
+            for attr in obj_dict
+            if not attr.startswith("_")
+        }
+
+    @backoff.on_exception(
+        backoff.expo,
+        (
+            braintree.exceptions.GatewayTimeoutError,
+            braintree.exceptions.RequestTimeoutError,
+            braintree.exceptions.ServerError,
+            braintree.exceptions.ServiceUnavailableError,
+            braintree.exceptions.TooManyRequestsError,
+        ),
+        max_tries=5,
+    )
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        start_date = stream_slice.get(self.cursor_field or "start_date")
+        items = self.get_items(start_date)
+        for item in items:
+            item = self.get_json_from_resource(item)
+            item = self.model(**item)
+            yield item.dict(exclude_unset=True)
```

Comment:
What does `exclude_unset=True` mean? Is it removes keys with None value? If yes, then why do we need it?

Reply:
It removes unset fields, i.e. one that are not required by model and havent been set explicitly when creating object from json
> Is it removes keys with None value? If yes, then why do we need it?
No it removes unset fields. If field was set when loading pydantic object from json it would be present in dict() output even if it has "None" value